pkg/sorter/sorter.go

```diff
@@ -0,0 +1,317 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package sorter
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/aws/amazon-ec2-instance-selector/v2/pkg/instancetypes"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/oliveagle/jsonpath"
+)
+
+const (
+	// Sort direction
+
+	sortAscending  = "ascending"
+	sortAsc        = "asc"
+	sortDescending = "descending"
+	sortDesc       = "desc"
+
+	// Not all fields can be reached through a json path (Ex: gpu count)
+	// so we have special flags for such cases.
+
+	gpuCountField              = "gpus"
+	inferenceAcceleratorsField = "inference-accelerators"
+)
+
+// sorterNode represents a sortable instance type which holds the value
+// to sort by instance sort
+type sorterNode struct {
+	instanceType *instancetypes.Details
+	fieldValue   reflect.Value
+}
+
+// Sorter is used to sort instance types based on a sorting field
+// and direction
+type Sorter struct {
+	sorters      []*sorterNode
+	sortField    string
+	isDescending bool
+}
+
+// NewSorter creates a new Sorter object to be used to sort the given instance types
+// based on the sorting field and direction
+//
+// sortField is a json path to a field in the instancetypes.Details struct which represents
+// the field to sort instance types by (Ex: ".MemoryInfo.SizeInMiB").
+//
+// sortDirection represents the direction to sort in. Valid options: "ascending", "asc", "descending", "desc".
+func NewSorter(instanceTypes []*instancetypes.Details, sortField string, sortDirection string) (*Sorter, error) {
```

Comment:
I don't think this pkg has a reason to construct objects. The preparation and sorting isn't really a two-step process. I think a simple semantic would be:

```
import github.com/aws/amazon-ec2-instance-selector/v2/pkg/sorter

instanceTypes, err := sorter.Sort(instanceTypes []*instancetypes.Details, sortField string, sortDirection string)
```

Reply:
That is a very good point. There would never be a reason to construct a sorter when you do not intend to sort. I have changed my sorting implementation to follow your suggestion and I have changed all sorting tests accordingly.